sdk/core/Azure.Core/src/Pipeline/BearerTokenAuthenticationPolicy.cs

```diff
@@ -289,58 +288,78 @@ public async ValueTask<string> GetHeaderValueAsync(HttpMessage message, TokenReq
+                {
+                    DateTimeOffset now = DateTimeOffset.UtcNow;
+                    if (localInfoTcs.Task.Status == TaskStatus.RanToCompletion && now < localInfoTcs.Task.Result.ExpiresOn && now < localInfoTcs.Task.Result.RefreshOn)
+                    {
+                        // localState entity has a valid token, no need to enter lock.
+                        return (localInfoTcs, default, false);
+                    }
+                }
                 lock (_syncObj)
                 {
+                    var (currentContext, infoTcs, backgroundUpdateTcs) = _state;
                     // Initial state. GetTaskCompletionSources has been called for the first time
-                    if (_infoTcs == null || RequestRequiresNewToken(context))
+                    if (infoTcs == null || RequestRequiresNewToken(context, currentContext))
                     {
-                        _currentContext = context;
-                        _infoTcs = new TaskCompletionSource<HeaderValueInfo>(TaskCreationOptions.RunContinuationsAsynchronously);
-                        _backgroundUpdateTcs = default;
-                        return (_infoTcs, _backgroundUpdateTcs, true);
+                        currentContext = context;
+                        infoTcs = new TaskCompletionSource<HeaderValueInfo>(TaskCreationOptions.RunContinuationsAsynchronously);
+                        backgroundUpdateTcs = default;
+                        _state = new TokenRequestState(currentContext, infoTcs, backgroundUpdateTcs);
+                        return (infoTcs, backgroundUpdateTcs, true);
                     }
 
                     // Getting new access token is in progress, wait for it
-                    if (!_infoTcs.Task.IsCompleted)
+                    if (!infoTcs.Task.IsCompleted)
                     {
-                        _backgroundUpdateTcs = default;
-                        return (_infoTcs, _backgroundUpdateTcs, false);
+                        // Only create new TokenRequestState if necessary.
+                        if (backgroundUpdateTcs != null)
+                        {
+                            backgroundUpdateTcs = default;
+                            _state = new TokenRequestState(currentContext, infoTcs, backgroundUpdateTcs);
+                        }
+                        return (infoTcs, backgroundUpdateTcs, false);
                     }
 
                     DateTimeOffset now = DateTimeOffset.UtcNow;
                     // Access token has been successfully acquired in background and it is not expired yet, use it instead of current one
-                    if (_backgroundUpdateTcs != null &&
-                        _backgroundUpdateTcs.Task.Status == TaskStatus.RanToCompletion &&
-                        _backgroundUpdateTcs.Task.Result.ExpiresOn > now)
+                    if (backgroundUpdateTcs != null &&
```

Comment:
Should we move these conditions to TokenRequestState and make them properties. Might make this method logic a bit simpler and remove the duplication with the fast path.

Reply:
Made them methods as they either take in the incoming TokenRequestContext or DateTimeOffset. We could make the ones taking DateTimeOffset into properties and each of the 3 checks would fetch their own DateTimeOffset.UtcNow. Let me know what you'd like me to do.